text/0005-firefox-data-stores.md

```diff
@@ -0,0 +1,21 @@
+
+# Problem
+
+There are a large number of data stores being used to hold data across all Firefox instances and platforms. There is currently nowhere that documents all these data stores and how they share information between them. We have 3 platforms, desktop, Android and iOS and each of these platforms shares data with the others. But not all the information is synced completely across all platforms, and the data itself is not stored in the same way everywhere. When starting to think about how we might provide a holistic user data solution for Mozilla, it is important to understand the scope of the problem and the specifics of the data that is being handled. 
+
+We therefore need to look at all of the data stores that we are currently using and document their purpose, structure, the kind of data that they hold and how they share that data.
+
+
+# Solution
+
+The extensive documentation can be found in the [Firefox Data Stores](https://github.com/mozilla/firefox-data-store-docs) repository.
+
+* There are a total of 46 separate data stores in Firefox for desktop. Within these 46 stores, there are 10 different storage formats. Very few of these stores are well documented in the code. 
+* Firefox for Android has only 3 stores as it consolidates many of the separate stores found in desktop Firefox into a single SQLite database. It uses only 3 different storage formats, SQLite, .json & .js. Android documents it's data stores well and uses a single file to define the schema for all it's tables.
```

Comment:
This isn't really true.

Firefox for Android uses Gecko's form history store, Gecko prefs, and the vast majority of the same files as desktop.

The differences are:

- It uses the SQLite logins store instead of the modern JSON one.
- It uses its own browser DB (and perhaps metadata DB) instead of `places.sqlite` and `favicons.sqlite`.
- It adds several Android XML preference files ("SharedPreferences").
- `times.json` and `signedInUser.json` are written by Java code in addition to, or instead of, Gecko code.
- It encapsulates most of its data handling code in a smaller set of places, typically behind a `ContentProvider` interface.
- It typically doesn't expose as many interfaces for editing Gecko-side data, so the Android codebase typically doesn't mention them.